File: net/TestVisualAssert/TestSoftVisualAssertCallStack.cs
using NUnit.Framework;
using System;

namespace Giis.Visualassert
{
	/// <summary>
	/// Checks the stack trace items shown in messages.
	/// WARNING: Expected messages (at the end of this class) are sensitive to 
	/// the line numbers where the asserts are thrown.
	/// Adjust the numbers if new lines are added to the code
	/// </summary>
	public class TestSoftVisualAssertCallStack
	{

		//Deep stack assert invocations
		public class OtherClass
		{
			public void DoAssert(SoftVisualAssert va)
			{
				va.AssertEquals("xy vw", "xy zz vw", "", "fstack12.html");
			}
		}

		private void CallDoAssert(SoftVisualAssert va)
		{
			OtherClass cls = new OtherClass();
			cls.DoAssert(va);
		}

		/*
		 * Conditions:
		 * -default(1)/more traces (0 already tested)
		 * -single/multiple fault
		 * -assert in the same test case/deeper in call stack (test class/other class)
		 */
		[Test]
		public void TestSingleStackItem()
		{
			SoftVisualAssert va = new SoftVisualAssert(); //default stack 1
			va.AssertEquals("ab zz cd", "ab cd", "", "fstack11.html");
			CallDoAssert(va);
			try
			{
				va.AssertAll();
				Assert.Fail("this should fail");
			}
			catch (Exception e)
			{
				/*System.Diagnostics.Debug.WriteLine(e.Message);*/ Assert.AreEqual(GetExpectedSingleStackItem(), e.Message.Replace("\\", "/"));
			}
		}
		[Test]
		public void TestMultipleStackItem()
		{
			SoftVisualAssert va = (SoftVisualAssert)new SoftVisualAssert().SetCallStackLength(3);
			CallDoAssert(va);
			try
			{
				va.AssertAll();
				Assert.Fail("this should fail");
			}
			catch (Exception e)
			{
				/*System.Diagnostics.Debug.WriteLine(e.Message);*/ Assert.AreEqual(GetExpectedMultipleStackItem(), e.Message.Replace("\\", "/"));
			}
		}

		private string GetExpectedSingleStackItem()
		{
			return "There are 2 failed assertion(s)\n"
					+ "Failure 1: Strings are different.\n"
					+ "- Visual diffs at: ../../../../reports/fstack11.html\n"
					+ "- Call Stack:\n"
					+ "    at Giis.Visualassert.TestSoftVisualAssertCallStack.TestSingleStackItem(TestSoftVisualAssertCallStack.cs:40)\n"
					+ "Failure 2: Strings are different.\n"
					+ "- Visual diffs at: ../../../../reports/fstack12.html\n"
					+ "- Call Stack:\n"
					+ "    at Giis.Visualassert.TestSoftVisualAssertCallStack.OtherClass.DoAssert(TestSoftVisualAssertCallStack.cs:20)";
		}

		private string GetExpectedMultipleStackItem()
		{
			return "There are 1 failed assertion(s)\n"
					+ "Failure 1: Strings are different.\n"
					+ "- Visual diffs at: ../../../../reports/fstack12.html\n"
					+ "- Call Stack:\n"
					+ "    at Giis.Visualassert.TestSoftVisualAssertCallStack.OtherClass.DoAssert(TestSoftVisualAssertCallStack.cs:20)\n"
					+ "    at Giis.Visualassert.TestSoftVisualAssertCallStack.CallDoAssert(TestSoftVisualAssertCallStack.cs:27)\n"
					+ "    at Giis.Visualassert.TestSoftVisualAssertCallStack.TestMultipleStackItem(TestSoftVisualAssertCallStack.cs:56)";
		}

	}
}

File: java/src/test/java/giis/visualassert/TestSoftVisualAssertCallStack.java
package giis.visualassert;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertThrows;

import org.junit.Test;

import giis.visualassert.portable.CallStack;
import giis.visualassert.portable.JavaCs;

/**
 * Checks the stack trace items shown in messages.
 * WARNING: Expected messages (at the end of this class) are sensitive to 
 * the line numbers where the asserts are thrown.
 * Adjust the numbers if new lines are added to the code
 */
public class TestSoftVisualAssertCallStack {
	
	private String innerClassSeparator = JavaCs.isJava() ? "$" : ".";

	//Deep stack assert invocations
	public class OtherClass {
		public void doAssert(SoftVisualAssert va) {
			va.assertEquals("xy vw", "xy zz vw", "", "fstack12.html");			
		}
	}

	private void callDoAssert(SoftVisualAssert va) {
		OtherClass cls=new OtherClass();
		cls.doAssert(va);
	}
	
	/*
	 * Conditions:
	 * -default(1)/more traces (0 already tested)
	 * -single/multiple fault
	 * -assert in the same test case/deeper in call stack (test class/other class)
	 */
	@Test
	public void testSingleStackItem() {
		SoftVisualAssert va = new SoftVisualAssert(); //default stack 1
		va.assertEquals("ab zz cd", "ab cd", "", "fstack11.html");
		callDoAssert(va);
		AssertionError e = assertThrows(AssertionError.class, () -> {
			va.assertAll();
		});
		assertEquals(CallStack.normalize(getExpectedSingleStackItem()).toLowerCase(), CallStack.normalize(e.getMessage()).toLowerCase());
	}
	@Test
	public void testMultipleStackItem() {
		SoftVisualAssert va = new SoftVisualAssert().setCallStackLength(3);
		callDoAssert(va);
		AssertionError e = assertThrows(AssertionError.class, () -> {
			va.assertAll();
		});
		assertEquals(CallStack.normalize(getExpectedMultipleStackItem()).toLowerCase(), CallStack.normalize(e.getMessage()).toLowerCase());
	}
	@Test
	public void testZeroStackItem() {
		SoftVisualAssert va = new SoftVisualAssert().setCallStackLength(0);
		va.assertEquals("ab zz cd", "ab cd", "", "fstack11.html");
		callDoAssert(va);
		AssertionError e = assertThrows(AssertionError.class, () -> {
			va.assertAll();
		});
		assertEquals(CallStack.normalize(getExpectedZeroStackItem()), CallStack.normalize(e.getMessage()));
	}
	
	private String getExpectedSingleStackItem() {
		return "There are 2 failed assertion(s)\n"
				+ "Failure 1: Strings are different. First diff at line 1 column 4.\n"
				+ "- Visual diffs at: target/fstack11.html\n"
				+ "- Call Stack:\n"
				+ "    at giis.visualassert.TestSoftVisualAssertCallStack.testSingleStackItem(TestSoftVisualAssertCallStack.java:37)\n"
				+ "Failure 2: Strings are different. First diff at line 1 column 4.\n"
				+ "- Visual diffs at: target/fstack12.html\n"
				+ "- Call Stack:\n"
				+ "    at giis.visualassert.TestSoftVisualAssertCallStack" + innerClassSeparator + "OtherClass.doAssert(TestSoftVisualAssertCallStack.java:19)";
	}

	private String getExpectedMultipleStackItem() {
		return "There are 1 failed assertion(s)\n"
				+ "Failure 1: Strings are different. First diff at line 1 column 4.\n"
				+ "- Visual diffs at: target/fstack12.html\n"
				+ "- Call Stack:\n"
				+ "    at giis.visualassert.TestSoftVisualAssertCallStack" + innerClassSeparator + "OtherClass.doAssert(TestSoftVisualAssertCallStack.java:19)\n"
				+ "    at giis.visualassert.TestSoftVisualAssertCallStack.callDoAssert(TestSoftVisualAssertCallStack.java:25)\n"
				+ "    at giis.visualassert.TestSoftVisualAssertCallStack.testMultipleStackItem(TestSoftVisualAssertCallStack.java:49)";
	}

	private String getExpectedZeroStackItem() {
		return "There are 2 failed assertion(s)\n"
				+ "Failure 1: Strings are different. First diff at line 1 column 4.\n"
				+ "- Visual diffs at: target/fstack11.html\n"
				+ "Failure 2: Strings are different. First diff at line 1 column 4.\n"
				+ "- Visual diffs at: target/fstack12.html";
	}
}
